proposals/explicit-api-mode.md

```diff
@@ -0,0 +1,154 @@
+# Explicit API mode
+
+* **Type**: Design proposal
+* **Author**: Ilya Gorbunov
+* **Contributors**: Leonid Startsev, Roman Elizarov
+* **Status**: Prototype
+* **Original proposal and discussion**: [KEEP-45](https://github.com/Kotlin/KEEP/issues/45)
+
+## Synopsis
+
+Provide an 'Explicit API' mode in the compiler which helps library authoring.
+Such mode should prevent delivery of unintended public API/ABI to the clients by requiring explicit visibility modifier and explicit return types for public declarations.
+
+## Motivation
+
+There were a couple of hot discussions about the default visibility before release:
+
+* https://discuss.kotlinlang.org/t/public-by-default-for-classes/110
+* https://discuss.kotlinlang.org/t/kotlins-default-visibility-should-be-internal/1400
+
+While such a decision is convenient for application development, the main concern against public-by-default visibility was that it becomes too easy for library authors to expose something accidentally, release it, and then have to make breaking changes to hide it back.
+
+## Proposal
+
+Introduce 'Explicit API' compiler mode. Compilation in such mode differs from the default mode in the following aspects:
+
+* Compiler requires you to specify explicit visibility for a declaration when leaving default visibility would result in exposing that declaration to the public API.
+
+* Compiler requires you to specify the explicit type of property/function when it is exposed to the public/published API.
+
+* Compiler requires you to explicitly [propagate experimental status](https://kotlinlang.org/docs/reference/experimental.html#propagating-use) for functions which contain experimental types in the signature.
+
+* Compiler warns you when exposed to public API declaration does not have a KDoc.
+
+### Public API definition
+
+#### Classes
+
+A class is considered to be effectively public if all of the following conditions are met:
+
+ - it has one of the following visibilities in Kotlin:
+    - no visibility (means no Kotlin declaration corresponds to this compiled class)
+    - *public*
+    - *protected*
+ - it isn't a local class
+ - it contains at least one effectively public member, in case if the class corresponds
+   to a kotlin *file* with top-level members or a *multifile facade*
```

Comment:
I don't see how the item on faced classes is relevant to the KEEP. We don't enforce "explicit public" or "explicit KDocs" on _file facades_ here. I suggest removing this item.